restart.c

```diff
@@ -0,0 +1,400 @@
+#include "memcached.h"
+
+#include "restart.h"
+
+#include <stdlib.h>
+#include <sys/mman.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <string.h>
+
+typedef struct _restart_data_cb restart_data_cb;
+
+struct _restart_data_cb {
+    void *data; // user supplied opaque data.
+    struct _restart_data_cb *next; // callbacks are ordered stack
+    restart_check_cb ccb;
+    restart_save_cb scb;
+    char tag[RESTART_TAG_MAXLEN];
+};
+
+// TODO: struct to hand back to caller.
+static int mmap_fd = 0;
+static void *mmap_base = NULL;
+static size_t slabmem_limit = 0;
+char *memory_file = NULL;
+
+static restart_data_cb *cb_stack = NULL;
+
+// Allows submodules and engines to have independent check and save metadata
+// routines for the restart code.
+void restart_register(const char *tag, restart_check_cb ccb, restart_save_cb scb, void *data) {
+    restart_data_cb *cb = calloc(1, sizeof(restart_data_cb));
+
+    // Handle first time call initialization inline so we don't need separate
+    // API call.
+    if (cb_stack == NULL) {
+        cb_stack = cb;
+    } else {
+        // Ensure we fire the callbacks in registration order.
+        // Someday I'll get a queue.h overhaul.
+        restart_data_cb *finder = cb_stack;
+        while (finder->next != NULL) {
+            finder = finder->next;
+        }
+        finder->next = cb;
+    }
+
+    // TODO: Write a safe_strncpy into util.c and stop using it raw.
+    // or a local strlcpy.
+    strncpy(cb->tag, tag, RESTART_TAG_MAXLEN);
```

Comment:
Yes! I love code reviews. I missed that TODO despite reading over the diff a dozen times. strncpy is important to avoid. I don't think it would cause a fault but it'll definitely cause ASAN to freak out. Will fix this after thinking about it a bit.

Reply:
fixed. wrote up a slow safe_strcpy which should avoid over-reading src, but still requires src to be null terminated properly. can also tell if it was truncated or not, but that function doesn't use that bit.